zerver/lib/events.py

```diff
@@ -277,6 +277,7 @@ def fetch_initial_state_data(
         # fit into that framework.
         realm_authentication_methods_dict = realm.authentication_methods_dict()
         state["realm_authentication_methods"] = realm_authentication_methods_dict
+        state["unavailable_authentication_methods"] = realm.unavailable_authentication_methods
```

Comment:
I decided to add this as a separate value, instead of expanding what we have in `realm_authentication_methods` because:

1. The `realm_authentication_methods` API is nice with its format `{"Google": True, "Email": Faise ...}`. Adding a nested dict structure here to do `{"AzureAD": {"enabled": False, "available": False}, ...}` makes it so much uglier, while being **only** relevant to Cloud realms. So it's introducing complication into the format that's completely irrelevant for all self-hosted users
2. Even if we're not worrying about backward-compatibility of this API, changing the output format of  `realm.authentication_methods_dict()` would change it for  `authentication_methods` in `/api/v1/server_settings` unless we write some extra code to fork these codepaths

Reply:
Probably worth an API design discussion. It might be that existing clients don't use `realm_authentication_methods` (just using the copy in `server_settings`), in which case we might want to migrate those to objects with a couple properties for better future extensibility.

We will likely want to do that migration for other reasons anyway, like having authentication methods only that a certain group can use.